### PR TITLE
[WIP] Migrate to before̬̜input͎

### LINKED ISSUE #11438: Migrate to beforeInput vol 2
## 📝 Provide a description of the new feature

An epic for revival of #7462 and further work.

*   Feature branch: [`ck/epic/11438-migrate-to-beforeinput`](https://github.com/ckeditor/ckeditor5/tree/ck/epic/11438-migrate-to-beforeinput)
*   Draft PR: #11456

### ToC

*   #11448
*   #11626
*   #11636
*   #11736
*   #12058 
*   ckeditor/ckeditor5#11904
*   Post-MVP: #12425 

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (engine, enter, typing): Typing and text composition (IME) support refactored to use `beforeInput` DOM events instead of DOM mutations and keystrokes. Closes #11438.

Other (typing): Introduced a new `insertText` event mapped from a subset of `beforeInput` events.

Other (typing): The `delete` event is now mapped from a subset of `beforeInput` events.

Other (enter): The `enter` event is now mapped from a subset of `beforeInput` events.

Fix (engine): Fixed conversion of the non-collapsed selection anchored between `<br>` elements (in Safari).

MINOR BREAKING CHANGE (clipboard): The `DataTransfer` class was moved to the engine package.

MINOR BREAKING CHANGE: The `input` command is now deprecated, you should use `insertText` command instead.

### CODE REVIEW

**Comment by dufipl** [comment]
@oleq,
`shift` + `enter` doesn't work correctly on Safari. It behaves just like pressed only `enter`. 
On Firefox and Chrome, it works fine.
It is a regression, on master it behaves correctly.

**Comment by Mgsy** [comment]
I love that finally CMD + backspace works :tada:. However, I've found one issue related to this feature.

### Firefox

1.  Put the caret at the end of a single-line paragraph.
2.  Press CMD + backspace.
3.  Repeat step 2.

The editor crashes.

```
Uncaught TypeError: selection.getFirstRange() is null
    execute deletecommand.js:121
    _runPendingChanges model.js:848
    enqueueChange model.js:253
    execute deletecommand.js:77
    decorate observablemixin.js:258
    fire emittermixin.js:199
    methodName observablemixin.js:262
    execute commandcollection.js:69
    execute editor.js:295
    init delete.js:72
    fire emittermixin.js:199
    fireListenerFor bubblingemittermixin.js:162
    fire bubblingemittermixin.js:92
    DeleteObserver deleteobserver.js:182
    fire emittermixin.js:199
    fireListenerFor bubblingemittermixin.js:162
    fire bubblingemittermixin.js:92
    fire domeventobserver.js:96
    onDomEvent inputobserver.js:58
    observe domeventobserver.js:79
    fire emittermixin.js:199
    domListener emittermixin.js:293
    attach emittermixin.js:219
    _addEventListener emittermixin.js:264
    addEventListener emittermixin.js:701
    listenTo emittermixin.js:117
    listenTo emittermixin.js:68
    listenTo emittermixin.js:65
    observe domeventobserver.js:77
    observe domeventobserver.js:76
    attachDomRoot view.js:287
    init classiceditorui.js:104
    create classiceditor.js:205
```

Works fine in Chrome/Safari.

**Comment by dufipl** [comment]
I've found an issue with Safari on iPad and it is a regression:
On iPad there are undo/redo buttons at the keyboard:
IMG_071859E45D09-1
On `master` they are not working when e.g. new line is added, but they are working when text is added.
On `ck/epic/11438-migrate-to-beforeinput` those buttons are not working anymore, even when text is being written.

**Comment by FilipTokarski** [comment]
### Android Gboard

1. Place caret at the end of a heading
2. Type

```js
Uncaught CKEditorError: model-nodelist-offset-out-of-bounds {"offset":19,"nodeList":[{"data":"Lists in the table"}]}
Read more: https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/error-codes.html#error-model-nodelist-offset-out-of-bounds
    at NodeList.offsetToIndex (nodelist.js:161:1)
    at Element.offsetToIndex (element.js:186:1)
    at getTextNodeAtPosition (position.js:1100:1)
    at Position.get textNode [as textNode] (position.js:218:1)
    at validateTextNodePosition (document.js:482:1)
    at Document._validateSelectionRange (document.js:391:1)
    at LiveSelection.<anonymous> (documentselection.js:666:1)
    at LiveSelection.fire (emittermixin.js:199:1)
    at LiveSelection._setRanges (selection.js:471:1)
    at LiveSelection.setTo (selection.js:370:1)
```


https://user-images.githubusercontent.com/34380544/159022138-c3a4c7fc-c98f-4fb6-8836-5fce43260212.mp4

**Comment by FilipTokarski** [comment]
### Android Swift keyboard

1. Select whole world
2. Change it to the one suggested by keyboard spellchecker

```js
Uncaught CKEditorError: insert-operation-position-invalid
Read more: https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/error-codes.html#error-insert-operation-position-invalid
    at InsertOperation._validate (insertoperation.js:117:1)
    at Model.on.priority (model.js:91:1)
    at Model.fire (emittermixin.js:199:1)
    at Model.<computed> [as applyOperation] (observablemixin.js:262:1)
    at Writer.insert (writer.js:219:1)
    at Insertion._insertPartialFragment (insertcontent.js:373:1)
    at Insertion.handleNodes (insertcontent.js:226:1)
    at insertcontent.js:76:1
    at Model.change (model.js:187:1)
    at insertContent (insertcontent.js:51:1)

Uncaught CKEditorError: model-nodelist-offset-out-of-bounds {"offset":11,"nodeList":[{"data":"Hello "}]}
Read more: https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/error-codes.html#error-model-nodelist-offset-out-of-bounds
    at NodeList.offsetToIndex (nodelist.js:161:1)
    at Element.offsetToIndex (element.js:186:1)
    at getTextNodeAtPosition (position.js:1100:1)
    at Position.get textNode [as textNode] (position.js:218:1)
    at validateTextNodePosition (document.js:482:1)
    at Document._validateSelectionRange (document.js:391:1)
    at LiveSelection.<anonymous> (documentselection.js:666:1)
    at LiveSelection.fire (emittermixin.js:199:1)
    at LiveSelection._setRanges (selection.js:471:1)
    at LiveSelection.setTo (selection.js:370:1)
```


https://user-images.githubusercontent.com/34380544/159026666-4b7a9257-489d-48f6-bc01-57296b23fa14.mp4

**Comment by FilipTokarski** [comment]
### Android GBoard

1. Press `enter` at the end of paragraph and type something

Empty paragraphs are created below, but selection is still in the same place. On `master` it's even more weird - you need to press `enter` twice to create a new paragraph, but then typing works fine.


https://user-images.githubusercontent.com/34380544/159028352-ddf2b68a-5b75-4359-9476-fddc1e39e2c0.mp4

**Comment by FilipTokarski** [comment]
## CHROME

### Steps

1. Select table cells that contain image (or other widget).
2. Press `backspace` or `delete`.

### Expected
All content from selected cells gets deleted:


https://user-images.githubusercontent.com/34380544/159230988-90e0af7b-5272-4271-b485-9b9bb50ea15f.mp4



### Actual
Nothing happens:


https://user-images.githubusercontent.com/34380544/159231010-42e7fd86-432d-4af3-81a7-d8ca14f03d9b.mp4

**Comment by dufipl** [comment]
The above ⬆️ issue reproduces also on Safari (on Firefox it works as expected).

**Comment by Reinmar** [comment]
FYI: the latest work is in branch `ck/11438-beforeinput-ime-research-vol1.1`.

Shouldn't this PR be closed?

**Comment by niegowski** [comment]
> Shouldn't this PR be closed?

No. This is a feature branch and those IME research branches are based on this feature branch.

File: packages/ckeditor5-engine/src/view/datatransfer.ts [hunk 1]
```diff
@@ -4,14 +4,20 @@
...
  */
 
+type DomDataTransfer = globalThis.DataTransfer;
+
 /**
  * A facade over the native [`DataTransfer`](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer) object.
  */
 export default class DataTransfer {
-	constructor( nativeDataTransfer ) {
+	public readonly files: File[];
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/datatransfer.ts]:
```suggestion
	public readonly files: Array<File>;
```

File: packages/ckeditor5-engine/src/view/view.ts [hunk 2]
```diff
@@ -214,11 +212,9 @@ export default class View extends Observable {
 		this.addObserver( FakeSelectionObserver );
 		this.addObserver( CompositionObserver );
 		this.addObserver( ArrowKeysObserver );
+		this.addObserver( InputObserver );
 		this.addObserver( TabObserver );
-
-		if ( env.isAndroid ) {
-			this.addObserver( InputObserver );
-		}
+		this.addObserver( InputObserver );
```
Comment (arkflpc) [inline packages/ckeditor5-engine/src/view/view.ts:217]: Twice?

File: packages/ckeditor5-engine/tests/view/view/view.js [hunk 3]
```diff
@@ -90,18 +90,9 @@ describe( 'view', () => {
...
-	it( 'should add InputObserver on Android devices', () => {
-		const oldEnvIsAndroid = env.isAndroid;
-		env.isAndroid = true;
-
-		const newView = new View( new StylesProcessor() );
-		expect( newView.getObserver( InputObserver ) ).to.be.instanceof( InputObserver );
-
-		env.isAndroid = oldEnvIsAndroid;
-		newView.destroy();
+		expect( view.getObserver( InputObserver ) ).to.be.instanceof( InputObserver );
```
Comment (arkflpc) [inline packages/ckeditor5-engine/tests/view/view/view.js:95]: Twice?

File: packages/ckeditor5-typing/tests/manual/beforeinput.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+## Typing using `beforeinput` event
+
+**NOTE**: Run this test with `--debug=typing` and check the "log CKE5 typing events" to see the detailed CKE5 events in the console.
+
+For inspiration on scenarios check out:
+
+* [Typing events - scenario](https://www.notion.so/Typing-events-scenarios-8d750fcdc8984e4fa5126e78b98e4047)
```
Comment (arkflpc) [inline packages/ckeditor5-typing/tests/manual/beforeinput.md:7]: Does it make sense to put in the public repo links to private Notion?

**Review by arkflpc — CHANGES_REQUESTED** [review]

**Review by arkflpc — APPROVED** [review]

**Comment by niegowski** [comment]
> MINOR BREAKING CHANGE: The `input` command is now deprecated, you should use `insertText` command instead.

To give some more background. This PR is changing the input flow, so the editor is now using the `beforeInput` events instead of DOM mutation-based input. The previous command (`input`) is still available as an alias and the same options should be accepted by both commands.

**Pull request merged**